server/protocols/protocols-pop3/src/test/java/org/apache/james/pop3server/POP3ServerTest.java

```diff
@@ -544,6 +549,55 @@ void testKnownUserInboxWithMessages() throws Exception {
         mailboxManager.deleteMailbox(mailboxPath, session);
     }
 
+    @Disabled("JAMES-3709 Concurrent deletes causes NPE when retrieving messages" +
+        "POP3: UIDL" +
+        "Delete the messages" +
+        "POP3: RETR XXX" +
+        "   /!\\ NPE")
+    @Test
+    void pop3SessionShouldTolerateConcurrentDeletes() throws Exception {
+        finishSetUp(pop3Configuration);
+
+        pop3Client = new POP3Client();
+        InetSocketAddress bindedAddress = new ProtocolServerUtils(pop3Server).retrieveBindedAddress();
+        pop3Client.connect(bindedAddress.getAddress().getHostAddress(), bindedAddress.getPort());
+
+        Username username = Username.of("foo2");
+        usersRepository.addUser(username, "bar2");
+
+        MailboxPath mailboxPath = MailboxPath.inbox(username);
+        MailboxSession session = mailboxManager.login(username, "bar2");
+
+        if (!mailboxManager.mailboxExists(mailboxPath, session).block()) {
+            mailboxManager.createMailbox(mailboxPath, session);
+        }
+
+        setupTestMails(session, mailboxManager.getMailbox(mailboxPath, session));
+
+        pop3Client.login("foo2", "bar2");
+        assertThat(pop3Client.getState()).isEqualTo(1);
+
+        POP3MessageInfo[] entries = pop3Client.listMessages();
+
+        assertThat(entries).isNotNull();
+        assertThat(entries.length).isEqualTo(2);
+        assertThat(pop3Client.getState()).isEqualTo(1);
+
+        mailboxManager.getMailbox(mailboxPath, session).delete(
+            Flux.from(mailboxManager.getMailbox(mailboxPath, session).listMessagesMetadata(MessageRange.all(), session))
+            .map(i -> i.getComposedMessageId().getUid())
+            .collectList().block(),
+            session);
+
+        Reader r = pop3Client.retrieveMessageTop(entries[0].number, 0);
+
+        assertThat(r).isNotNull();
+        assertThat(IOUtils.toString(r)).isEqualTo("-ERR\r\n");
+
```

Comment:
In order for the test to pass, after the fix, the two lines above should be:
assertThat(r).isNull();
which pop3client returns in case of an error. But this is what we want, since a fixed handler should respond with "-ERR message is already deleted".

(The two lines do work without the fix, since initially you get "+OK message follows" but with a null content stream, so the subsequent read via IOUtils.toString(r) cause the NPE. Then the "content" read is the resulting "-ERR".)

Reply:
Subttle bug!